crates/pathfinder/src/consensus/inner/p2p_task_tests.rs

```diff
@@ -0,0 +1,1402 @@
+        tokio::time::sleep(Duration::from_millis(300)).await;
+        env.verify_task_alive().await;
+
+        // Step 5: Send TransactionsFin with count=7 (should trigger rollback from 10 to
+        // 7)
+        env.p2p_tx
+            .send(Event::Proposal(
+                height_and_round,
+                ProposalPart::TransactionsFin(p2p_proto::consensus::TransactionsFin {
+                    executed_transaction_count: 7,
+                }),
+            ))
+            .expect("Failed to send TransactionsFin");
+        tokio::time::sleep(Duration::from_millis(300)).await;
+        env.verify_task_alive().await;
+
+        // Step 6: Send ProposalCommitment
+        env.p2p_tx
+            .send(Event::Proposal(
+                height_and_round,
+                create_proposal_commitment_part(2, proposal_commitment),
+            ))
+            .expect("Failed to send ProposalCommitment");
+        tokio::time::sleep(Duration::from_millis(300)).await;
+        env.verify_task_alive().await;
+
+        // Step 7: Send ProposalFin
+        env.p2p_tx
+            .send(Event::Proposal(
+                height_and_round,
+                ProposalPart::Fin(p2p_proto::consensus::ProposalFin {
+                    proposal_commitment: p2p_proto::common::Hash(proposal_commitment.0),
+                }),
+            ))
+            .expect("Failed to send ProposalFin");
+        tokio::time::sleep(Duration::from_millis(500)).await;
+
+        // Verify: Proposal event should be sent (rollback completed successfully)
+        // NOTE: We verify that a proposal event is sent, which indicates rollback
+        // completed. However, we cannot directly verify the transaction count
+        // in e2e tests because the validator is internal to p2p_task. The
+        // rollback logic itself is verified in unit tests (batch_execution.
+        // rs::test_transactions_fin_rollback). This e2e test verifies
+        // that rollback doesn't break the proposal flow end-to-end.
+        let proposal_cmd = wait_for_proposal_event(&mut env.rx_from_p2p, Duration::from_secs(2))
+            .await
+            .expect("Expected proposal event after ProposalFin");
+        verify_proposal_event(proposal_cmd, 2, proposal_commitment);
+
+        // Verify proposal parts persisted
+        // Query with validator_address (receiver) to get foreign proposals
+        verify_proposal_parts_persisted(&env.consensus_storage, 2, 1, &validator_address, 4);
+
+        // Note: Persisted proposal parts contain the original transactions (10), not
+        // the rolled-back count (7). Rollback happens in the validator's
+        // execution state at runtime, but the persisted parts reflect what was
+        // received from the network. The rollback verification (that execution
+        // state has 7 transactions) is covered in unit tests. Here we verify
+        // that all original batches are persisted.
```

Comment:
I'm not sure I understood this correctly. If there still are 10 persisted parts which have not been purged yet, there is always a risk of the process being restarted by the operator and recovery from parts persistence will yield an incorrect result - right?

My comment does not refer to the test being bad or anything similar. I'm thinking recovery from parts  should take this (ie rollback) into account.

Reply:
This essentially means that if the process crashes and recovers from persisted parts, it would restore 10 transactions instead of the rolled-back count of 7.

Your concern is valid. I'll add a comment noting that the recovery logic should take `TransactionsFin` into account to ensure the correct transaction count is restored after rollback, despite this not being the object of this test.

Reply:
https://github.com/eqlabs/pathfinder/pull/3104/commits/dcc8cdaf5915f5be02ad4d78442c21736f716180#diff-48d97864be7dfcfe14389b16a6f6b830f2c855a242af4bf111d28be412c8ce44R1047-R1050